File: Problems/Leetcode/2001-3000/2801-2900/2844.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minimumOperations(string num) {
        int n = num.size();

        int ans = n;
        for (int i = n - 1; i >= 0; --i) {
            if (num[i] == '0') {
                ans = min(ans, n - 1);
                for (int j = i - 1; j >= 0; --j) {
                    if (num[j] == '0' || num[j] == '5') {
                        ans = min(ans, n - j - 2);
                        break;
                    }
                }
            }
            if (num[i] == '5') {
                for (int j = i - 1; j >= 0; --j) {
                    if (num[j] == '2' || num[j] == '7') {
                        ans = min(ans, n - j - 2);
                        break;
                    }
                }
            }
        }
        return ans;
    }
};

int main() {
    Solution sol;

    return 0;
}

File: Problems/Leetcode/2001-3000/2801-2900/2844.py
from functools import cache


class Solution_1:
    def minimumOperations(self, num: str) -> int:
        n = len(num)

        @cache
        def dfs(i, cur) -> int:
            if i == n:
                return n if cur else 0
            r1 = dfs(i + 1, (cur * 10 + int(num[i])) % 25)
            r2 = dfs(i + 1, cur % 25) + 1
            return min(r1, r2)

        return dfs(0, 0)


class Solution_2:
    def minimumOperations(self, num: str) -> int:
        d = {"0": "05", "5": "27"}
        n = len(num)
        ans = n - 1 if "0" in num else n
        for i in range(n - 1, -1, -1):
            if num[i] in d:
                for j in range(i - 1, -1, -1):
                    if num[j] in d[num[i]]:
                        ans = min(ans, n - j - 2)
        return ans
